Cartesian chart

ColinEberhardt opened the pull request:
I think we need a more versatile chart component with axis labels and configurable scales. This is a quick prototype to test out a few ideas.

The non-standard part of this component is the way the scales work. In order to expose them as properties the settings from previous scales need to be re-applied and the new scale properties rebound.

What do others think - does this look viable?

chrisprice commented:
I know this isn't a problem specifically with this PR (and I don't have any ideas how to avoid it) but there's so much boilerplate for building static DOM structure.

Also, the re-binding stuff does look a bit odd. As they're unlikely to be changed, could they be passed in as arguments to the factory method?

ColinEberhardt commented:
You're right about the DOM construction code, it is very hard to determine what
the component creates from inspecting the code. I'll see if I can create a more
generic way of creating 'static' DOM from a structured model.
Not sure what you mean about passing the rebounding arguments to the factory
method? Could you give a quick example?
Thanks
Colin E.

On Wed, Aug 26, 2015 at 4:13 pm, ScottLogic/d3fc
<reply@reply.github.com>
wrote:
I know this isn't a problem specifically with this PR (and I don't have any
ideas how to avoid it) but there's so much boilerplate for building static DOM
structure.

Also, the re-binding stuff does look a bit odd. As they're unlikely to be
changed, could they be passed in as arguments to the factory method?

—
Reply to this email directly or view it on GitHub
[https://github.com/ScottLogic/d3fc/pull/518#issuecomment-135057024] .[https://github.com/notifications/beacon/ABDBfn09zHibsVueefFZHxVN9Dd1dZLtks5orc8cgaJpZM4FyU_b.gif]

chrisprice commented:
Sorry, that was a wordy was of saying -

```js
fc.chart.cartesianChart = function(xScale, yScale) {
```

And not letting them be subsequently changed.

ColinEberhardt commented:
> I know this isn't a problem specifically with this PR (and I don't have any ideas how to avoid it) but there's so much boilerplate for building static DOM structure.

You've got a point - I find ti hard to follow the DOM construction code. I've pushed a commit with a very different approach which uses `html` to 'bulk' add the entire structure then uses selectors to locate and update the various layout properties. 

https://github.com/ColinEberhardt/d3-financial-components/blob/86e1397f88f3935404dd66f5c9cb34c10073a867/src/chart/cartesianChart.js#L58-L82

I'm not so keen on having the scales provided as factory method args, it's not a pattern we use elsewhere.

chrisprice commented:
Is that not because we've not had a good use case?

It is used in other parts of d3 e.g.
https://github.com/mbostock/d3/wiki/Requests#d3_xhr.

On Tue, 1 Sep 2015 at 08:50 Colin Eberhardt <notifications@github.com>
wrote:

> I know this isn't a problem specifically with this PR (and I don't have
> any ideas how to avoid it) but there's so much boilerplate for building
> static DOM structure.
>
> You've got a point - I find ti hard to follow the DOM construction code.
> I've pushed a commit with a very different approach which uses html to
> 'bulk' add the entire structure then uses selectors to locate and update
> the various layout properties.
>
>
> https://github.com/ColinEberhardt/d3-financial-components/blob/86e1397f88f3935404dd66f5c9cb34c10073a867/src/chart/cartesianChart.js#L58-L82
>
> I'm not so keen on having the scales provided as factory method args, it's
> not a pattern we use elsewhere.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ScottLogic/d3fc/pull/518#issuecomment-136618737>.
>

ColinEberhardt commented:
Still not convinced about passing the scales to the factory function - it works, but the nested fluent syntax looks a bit ugly:

    var chart = fc.chart.cartesianChart(
            d3.scale.linear()
                .domain(fc.util.extent(data, 'x')),
            d3.scale.log()
                .domain(fc.util.extent(data, 'y')))
        .yOrient(chartConfig[0].value)
        .yAxisLabel(chartConfig[1].value)
        .xAxisLabel(chartConfig[2].value)
        .xAxisLabelHeight(Number(chartConfig[3].value))
        .yAxisLabelWidth(Number(chartConfig[4].value))
        .xAxisHeight(Number(chartConfig[5].value))
        .yAxisWidth(Number(chartConfig[6].value));

chrisprice commented:
Yea that does look odd. I'd imagined the scales would still have their
properties rebound though, would that help?

On Sat, 5 Sep 2015 21:27 Colin Eberhardt <notifications@github.com> wrote:

> Still not convinced about passing the scales to the factory function - it
> works, but the nested fluent syntax looks a bit ugly:
>
> var chart = fc.chart.cartesianChart(
>         d3.scale.linear()
>             .domain(fc.util.extent(data, 'x')),
>         d3.scale.log()
>             .domain(fc.util.extent(data, 'y')))
>     .yOrient(chartConfig[0].value)
>     .yAxisLabel(chartConfig[1].value)
>     .xAxisLabel(chartConfig[2].value)
>     .xAxisLabelHeight(Number(chartConfig[3].value))
>     .yAxisLabelWidth(Number(chartConfig[4].value))
>     .xAxisHeight(Number(chartConfig[5].value))
>     .yAxisWidth(Number(chartConfig[6].value));
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ScottLogic/d3fc/pull/518#issuecomment-137992921>.
>

ColinEberhardt commented:
@chrisprice I think this one is ready for a proper review, I've dropped the [WIP] prefix! Squashed, and rebased.

> I'd imagined the scales would still have their properties rebound though, would that help?

Yeah, that did, the API works quite well now.

I was going to change `linearTimeSeries` to use the same construction pattern, but realised it requires constructing the SVG container. We could probably do with debating this one a little more!

chrisprice commented on src/chart/cartesianChart.js:
```diff
@@ -0,0 +1,232 @@
+            titleHeight = 20,
+            yAxisLabel = 'y-axis',
+            xAxisLabel = 'x-axis',
+            title = 'Chart Title',
```
`chartLabel` for consistency? Also make the default value `chart`?

ColinEberhardt replied:
Good idea, I've also renamed `titleHeight` to `chartLabelHeight` for consistency.

chrisprice commented on src/util/rebind.js:
```diff
@@ -24,4 +24,19 @@
     };
 
+    function capitalizeFirstLetter(str) {
+        return str.charAt(0).toUpperCase() + str.slice(1);
```
`str[0].toUpperCase() + str.slice(1)`

chrisprice commented on src/util/rebind.js:
```diff
@@ -24,4 +24,19 @@
+    /**
+     * Rebinds multiple properties from the source component, adding the given prefix.
+     */
+    fc.util.rebindProperties = function(target, source, properties, prefix) {
```
How about calling it `rebindAll`, dropping `properties` and treating the arguments after `prefix` as excludes?

ColinEberhardt replied:
Nice idea - I'll give that a go.

chrisprice commented on src/chart/chart.css:
```diff
@@ -19,4 +19,15 @@
+    text-anchor: middle;
+}
+
+.cartesianChart {
```
Should we have the rule that all classes are dash-separated rather than camel-cased?

ColinEberhardt replied:
I think we already have that convention, I just ignored it!

ColinEberhardt commented:
Thanks for the review - `rebindAll` is a nice idea, and also forced me to put a bit more thought into which properties are actually rebound.

chrisprice commented:
Nice one, I'll take another look next time I'm in front of a computer

On Tue, 8 Sep 2015 23:28 Colin Eberhardt <notifications@github.com> wrote:

> Thanks for the review - rebindAll is a nice idea, and also forced me to
> put a bit more thought into which properties are actually rebound.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ScottLogic/d3fc/pull/518#issuecomment-138707591>.
>

chrisprice commented on src/chart/cartesianChart.js:
```diff
@@ -0,0 +1,233 @@
+                    </g>');
+
+                // configure the y axis containers based on y-axis orientation
+                var opposingOrientation = yAxis.orient() === 'left' ? 'right' : 'left';
```
There are a few statically defined values in the `layout` argument object. I assume that's because `layout` doesn't support applying deltas using the object syntax? Is that something we should create an issue for?

chrisprice commented on src/chart/cartesianChart.js:
```diff
@@ -0,0 +1,233 @@
+
+        // exclude scale because this component associates the scale with the axis, and
+        // exclude x axis orientation, as only y axis can be repositions to the left / right
+        fc.util.rebindAll(cartesianChart, xAxis, 'x', ['scale', 'orient']);
```
Is it worth allowing the x-axis to be positioned at the top?

ColinEberhardt replied:
Yes it would, although I think I've got a better idea for how to handle axis orientation, I'll just try out this alternative first.

chrisprice commented on src/util/rebind.js:
```diff
@@ -24,4 +24,30 @@
+     * optional exclusions parameter can be used to specify methods which should not
+     * be rebound.
+     */
+    fc.util.rebindAll = function(target, source, prefix, exclusions) {
```
Could the exclusions be made an array or var-args? I can see valuable use cases for both.

ColinEberhardt replied:
Nice idea - done.

chrisprice commented on src/util/rebind.js:
```diff
@@ -24,4 +24,30 @@
+        var bindings = {};
+        for (var property in source) {
+            if (source.hasOwnProperty(property) && exclusions.indexOf(property) === -1) {
+                bindings[prefix + capitalizeFirstLetter(property)] = property;
```
Would it easier to just call `d3.rebind` here?

ColinEberhardt replied:
I don't think that will work, these bindings need to be 'aliased' to a different property name.

ColinEberhardt commented:
@chrisprice I've made quite a big update to this code:

 + Axes are not `fc.series.axis` allowing the baseline to be configured
 + The chart layout has been completely changed and simplified. It is simple a single container (plot area) with a configurable margin. Axes are now on the chart boundaries (as you suggested ;-)
 + Both axes support 'orient'

This feels like a much more elegant approach.

chrisprice commented on src/chart/cartesianChart.js:
```diff
@@ -0,0 +1,211 @@
+                left: 0,
+                right: 40
+            },
+            yAxisLabel = 'y-axis',
```
I can't see anyone leaving these labels as-is, is it worth picking defaults that are more likely to be applicable? E.g. `''`?

ColinEberhardt replied:
done

chrisprice commented on src/chart/cartesianChart.js:
```diff
@@ -0,0 +1,211 @@
+                    alignItems: 'center',
+                    flexDirection: 'row'
+                };
+                if (yAxis.orient() === 'right') {
```
`yAxisLayout[yAxis.orient] = 0;`

ColinEberhardt replied:
done

chrisprice commented on src/chart/cartesianChart.js:
```diff
@@ -0,0 +1,211 @@
+            plotArea = x;
+            return cartesianChart;
+        };
+        cartesianChart.xAxisLabel = function(x) {
```
`xLabel`

ColinEberhardt replied:
done

chrisprice commented on src/chart/cartesianChart.js:
```diff
@@ -0,0 +1,211 @@
+            plotAreaMargin = x;
+            return cartesianChart;
+        };
+        cartesianChart.yAxisLabel = function(x) {
```
`yLabel`

ColinEberhardt replied:
done

chrisprice commented on src/chart/cartesianChart.js:
```diff
@@ -0,0 +1,211 @@
+            xAxisLabel = x;
+            return cartesianChart;
+        };
+        cartesianChart.plotAreaMargin = function(x) {
```
Just `margin`?

chrisprice commented on src/chart/cartesianChart.js:
```diff
@@ -0,0 +1,211 @@
+                var svg = containerDataJoin(container, [data]);
+                svg.enter().html(
+                    '<g class="title"> \
+                        <g layout-css="height: 0; width: 0"> \
```
I don't think this guarantees that labels are wholly within the bounding box?

ColinEberhardt commented:
@chrisprice - all review comments addressed, I think

chrisprice commented:
:+1: from me, if you feel inclined you could make that change but otherwise feel free to create an issue and merge.

ColinEberhardt commented:
Yay, time for a merge beer

chrisprice commented:
Going to push this out as 1.1.0, there's been a couple of things added
recently which are useful.

On Fri, 11 Sep 2015 at 17:47 Colin Eberhardt <notifications@github.com>
wrote:

> Yay, time for a merge beer
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ScottLogic/d3fc/pull/518#issuecomment-139595903>.
>

ColinEberhardt commented:
Cool, I'll add the docs page for the website shortly.

On Fri, Sep 11, 2015 at 5:48 pm, ScottLogic/d3fc
<reply@reply.github.com>
wrote:
Going to push this out as 1.1.0, there's been a couple of things added
recently which are useful.

On Fri, 11 Sep 2015 at 17:47 Colin Eberhardt <notifications@github.com>
wrote:

> Yay, time for a merge beer
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ScottLogic/d3fc/pull/518#issuecomment-139595903>.
>
—
Reply to this email directly or view it on GitHub
[https://github.com/ScottLogic/d3fc/pull/518#issuecomment-139596209] .[https://github.com/notifications/beacon/ABDBfkNOZ4AEmktSf8PZ8MdkPm_UlaNvks5owv1ngaJpZM4FyU_b.gif]
